### ISSUE TITLE
destBasePath can't be set

### ISSUE THREAD

**Issue by macrozone** [amwmedia/plop#106]
altough its used in the code, there is no way to define destBasePath in the plopfile

So if you provide a plop-pack which creates a certain folder structure, you have to introduce your own config to do so, you can't use config.destBasePath

**Comment by amwmedia** [comment]
the destination base path is the directory where the plopfile is found. This property can be set if you are using node-plop, but it is used internally to tell other plopfiles that have been loaded via `plop.load()` where to put their files.

**Comment by macrozone** [comment]
@amwmedia yeah, saw that to, but that does not solve or answer the issue.

imaging you develop a plop-pack that defines generators that create a certain folder structure, e.g. /src/components, /src/containers, /src/actions, etc.

now, you want to enable your users to change /src to something else, or to nest it further.

for that, you would need to define the destBasePath.

**Comment by amwmedia** [comment]
AH! ok, now I understand the request. for a plop-pack, you would specify an attribute on the second parameter of the `plop.load()` method ([config](https://github.com/amwmedia/plop/blob/master/plop-load.md#config)). This is your object that consumers can use to configure your plop-pack.

I hope that helps.

**Comment by macrozone** [comment]
@amwmedia yes, that would be the workaround. But i thought destBasePath would be exactly for this us e case. Now i need to implement a custom version of it...

**Comment by amwmedia** [comment]
`destBasePath` is used internally to map `plop.load`ed generators to the same basic root folder as their consumer. Prior to `plop.load` the generators in an npm module would need to define their action's destination paths with enough `../../` pathing to get them out of `node_modules` and back to the consuming codebase.

the config object isn't really a workaround, it is there for this purpose. It is a single place where the configuration of the plop-pack should happen.

**Comment by macrozone** [comment]
so destBasePath should be renamed to someting like plopFilePath ? because it clearly says: this is the folder where every file will get created. And that's a folder the user clearly want to be able to set.

I tried to set it on `plop.load("myploppack", {destBasePath: "./src"})` but this does not work

**Comment by amwmedia** [comment]
`plopfilePath` is actually another thing altogether. The complexity here is that a `load`ed generator should find a template locally (in node_modules), but drop its generated code into the `plopfilePath` of the consuming plopfile.

**Comment by macrozone** [comment]
@amwmedia yes, but that's not what i meant.

There needs to be multiple path that have nothing to do with each other:

- plopfile path: is the path to the plopfile inside the project (at the root of the project)
- destBasePath: the path where files will be created. Should be customizable by the project-team (via plopfile)
- templatePath (?): should either point to the template directory inside of the project and maybe to the template directory in the ploppack / npm module

**Comment by amwmedia** [comment]
For simplicity, plop take the approach that everything is based on the plopfile location. All destination paths are relative to the plopfile location, all templates are located relative to the plopfile location, etc. This allows you to execute the `plop` command from anywhere inside the project and have files created in the correct places (the current working directory doesn't matter). The only reason that the `destBasePath` is needed is for `plop.load()` where you have code that is running from `node_modules` but should be dumping its files into another location that is out of its control (the consuming plopfile path). 

If you want to allow the consuming plopfile to tell your plop-pack where to drop its files, you can accept a value in the config options that you can then use in your path definitions. This becomes just another config value for your plop-pack module, there's no special meaning to it, and you can call it whatever you like.

**Comment by tarikhamilton** [comment]
@macrozone Still use plop? I used this approach to create a global plopfile, if that is what you were essentially trying to achieve.

I used this `path` in the Action. While `__dirname` is the plopfile location, `.` will still give you the relative path that you called it from!

`path: path.join(path.relative(__dirname, '.'), 'src/components/{{name}}/index.js'),`

@amwmedia I haven't had time to fully digest `plop.load()`. I personally wanted a global plop that I could use without having to add plopfiles to a repo. Could that still be achieved with `plop.load()`? It seems like I'd need it to be a node module, which means adding it as a dependency, which is what I wanted to avoid. Thanks for the great project.
